client/client.go

```diff
@@ -859,7 +859,9 @@ func (c *Client) Shutdown() error {
 			arGroup.AddCh(ar.DestroyCh())
 		}
 	} else {
-		// In normal mode call shutdown
+		// In normal mode optionally drain the node before shutting down any
+		// remaining alloc runners.
+		c.DrainSelf()
```

Comment:
As we discussed offline while I think `cancel the drain` will be a great escape hatch to document for users, I'm not sure it's sufficient for the inplace Nomad upgrade case.

I'm assuming there are 2 common reasons for shutting down a Nomad agent:
1. Upgrading the `nomad` binary
2. Decommissioning a node

(I think there are other reasons like "something is broken and turning things off and on again is always worth a shot" -- but _hopefully_ that's far less common than the 2 above.)

This behavior suits `#2` perfectly, and manually cancelling a drain to speed up `#2` is totally reasonable.

However for `#1` (upgrading Nomad), not only does it require an extra command/API-call (and the error handling considerations that goes with any operation), but it also incurs 2 useless Raft writes per Node (start the drain / stop the drain). For large clusters looking to roll clients as quickly as possible that could be kind of annoying as restarts likely already increase Leader load when they re-register and sync allocs on startup.

So it'd be nice to tell users `When decommissioning use SIGINT and when upgrading use SIGTERM` or similar.

Windows will require an API (and associated `nomad node shutdown ...` command which would be pretty cool), but I think we can leave that to a future PR.

Reply:
Aye, I can get something wired into the signal handler for this.

Reply:
@schmichael I dug into this and remembered that we've got an existing facility for graceful exit with [`leave_on_interrupt`](https://developer.hashicorp.com/nomad/docs/configuration#leave_on_interrupt) and [`leave_on_terminate`](https://developer.hashicorp.com/nomad/docs/configuration#leave_on_terminate) on either SIGINT or SIGTERM, which for clients does not actually do anything today because it ends up [here](https://github.com/hashicorp/nomad/blob/main/client/client.go#L759-L763
):

```go
// Leave is used to prepare the client to leave the cluster
func (c *Client) Leave() error {
	// TODO
	return nil
}
```

So I think we can move the call to `DrainSelf` into the `Leave` method, and that'll let the existing `leave_on_interrupt` and `leave_on_terminate` configurations actually do something for clients. Cluster adminstrators get their pick of which signal they want to use for graceful exits with drain, and if neither are set we don't drain at all.

Reply:
That `TODO` is from 2015! Finally done! :grinning:

Reply:
:facepalm: Wow I had totally forgotten about those flags. Fantastic work.